### PR TITLE
Add site logo crop

### LINKED ISSUE #29217: Site Logo Block: Add the same dimensions settings as the Image block
## Description

The Site Logo block only has a "Image width" setting which not only can malfunction, it's also different than the dimensions settings of the Image block.

Since both blocks are used to render an image, they should have the same settings available.
For example, the Site Logo block should probably use the same `ImageSizeControl` component to handle its dimensions as the Image block.

## Step-by-step reproduction instructions

1. Insert a Site Logo block (add an image if needed), followed by a Paragraph block. 
2. Resize the logo using the resize handlers around the image itself.
3. Then, try resizing the logo using the slider in the block settings sidebar.

## Expected behaviour

The Site Logo wrapper resized along with the image.

## Actual behaviour

The Site Logo wrapper didn't resize appropriately, and the image was either overlapping the following block, or smaller than the wrapper.

## Screenshots or screen recording (optional)

| Original | Resize to Small | Resize to Large |
| - | - | - |
| Screenshot 2021-02-22 at 12 11 41 | Screenshot 2021-02-22 at 12 11 17 | Screenshot 2021-02-22 at 12 11 27 |

## WordPress information
- WordPress version: 5.7-beta2-50343
- Gutenberg version: 10.0.0-rc.1
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

## Device information
- Device: Desktop
- Operating system: macOS 11.2.1
- Browser: Chrome 88.0

### PR DESCRIPTION
**Pull request by ajlende**
## Description

Part of #30406.

Also fixes #29217 in 93dc9675cb448586621a840a310202fe3acdee65. This can be moved to another PR if people would prefer to keep the fix separate.

Adds cropping to the site logo by:

1. Moving the cropping component to the @wordpress/block-editor package and exporting it as an `__experimental` component
2. Using the exported cropper in the SiteLogo block

## How has this been tested?

1. Add SiteLogo block.
2. See that toolbar doesn't appear when the logo is empty.
3. Apply a crop when a logo is present.
4. See that the site logo is updated when the crop is applied.

## Screenshots 

image

image

crop-logo

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
This is going to be so extremely useful for the site logo block, thanks for working on it! (Oh and by the way, I feel like at this point you've deserved to be admitted as a contributor, if you'd like to work on the main branch rather than your fork 😉). This is what I see:

site logo crop

In general this seems to be working well. There are some flow aspects to the cropping system that could be refined to benefit both this and the Image block, but we can look at that separately.

As shown at the end of the GIF, if I change the aspect ratio of the Site Logo the general shape/silhouette of the overall block does not appear to be updated. I wonder if that's related to #29217?

**Comment by ajlende** [comment]
> I wonder if that's related to #29217?

It seems like it might be. The width calculation seems to be a render behind which causes issues both here and when the resize happens.

**Comment by ajlende** [comment]
@jasmussen I finally circled back around on this, and I think the issue is fixed now after the rebase. Give it another look, and if everything is good we can get this merged :)

EDIT: I was able to reproduce it again now. Looking into the fix.

**Comment by jasmussen** [comment]
I was just about to compile this, then saw the edit :D 

Please do ping me as soon as you can use my help, whether as a review or otherwise.

**Review by jasmussen — APPROVED** [review]
Very nice, thank you!

site logo
ar

If you can get a sanity check on the code it would be nice, but functionally this is solid!

File: packages/block-editor/README.md [hunk 1]
```diff
@@ -423,6 +423,14 @@ _Returns_
...
 -   `string`: Gradient value.
 
+<a name="ImageEditingProvider" href="#ImageEditingProvider">#</a> **ImageEditingProvider**
+
+Undocumented declaration.
+
+<a name="ImageEditor" href="#ImageEditor">#</a> **ImageEditor**
+
+Undocumented declaration.
+
```
Comment (ajlende) [inline packages/block-editor/README.md]: @ellatrix you were doing some work on the image editor previously. Do you have an opinion on making the image editing tools part of the block-editor package? I think that the public API could be simplified to not need the provider as a separate component. I could do that here or mark it as experimental and do that in a follow-up if you think that would be better.

Comment (ajlende) [reply packages/block-editor/README.md]: #23155 was asking for the image editor to be exported for other blocks too.

**Comment by jasmussen** [comment]
Is this one blocked?

**Comment by ajlende** [comment]
There was a bug with the applied image not showing up in trunk for a while, but it seems to be resolved. I merged in the changes and it seems to be working as intended again.

File: packages/block-editor/README.md (hunk 1 above)
Comment (ajlende) [reply packages/block-editor/README.md]: I went ahead and marked it as experimental in 28ba223

File: packages/block-library/src/site-logo/edit.js [hunk 2]
```diff
@@ -179,6 +185,63 @@ const SiteLogo = ( {
...
+					naturalHeight={ naturalHeight }
+					naturalWidth={ naturalWidth }
+				/>
+			</ImageEditingProvider>
+		) : (
+			<ResizableBox
+				size={ {
+					width: currentWidth,
+					height: currentHeight,
+				} }
```
Comment (ajlende) [inline packages/block-library/src/site-logo/edit.js:225]: Using `currentWidth` and `currentHeight` fixes the bug described in #29217

**Review by vcanales — COMMENTED** [review]
When cropping, if we deselect the image block the crop tool stays on. This creates a couple of issues;

1. At first it's hard to tell that the crop hasn't been applied, even after updating the post. I think it might be good to either Apply or Cancel when deselecting the block, or Applying when the post is saved while cropping.

Kapture 2021-09-14 at 10 01 53

2. It becomes hard to select the image block again; dragging seems to interfere with clicking on the block. A way I found to achieve it was to tab into the block, and then clicking it.

Kapture 2021-09-14 at 10 05 58

I'm also wondering if it is possible to allow to "decrop" the image; as of right now, Applying the crop creates a new one, and in order to un-do from this point it's necessary to replace the cropped image with the original one.

**Comment by ajlende** [comment]
> 1. At first it's hard to tell that the crop hasn't been applied, even after updating the post. I think it might be good to either Apply or Cancel when deselecting the block, or Applying when the post is saved while cropping.
>
> 2. It becomes hard to select the image block again; dragging seems to interfere with clicking on the block. A way I found to achieve it was to tab into the block, and then clicking it.

Nice catch! Deselecting should match the behavior of the image block now.

> I'm also wondering if it is possible to allow to "decrop" the image; as of right now, Applying the crop creates a new one, and in order to un-do from this point it's necessary to replace the cropped image with the original one.

Seems like a good idea—we could update the cropper to make sure an undo layer is created when hitting apply so we can, at least, hit ctrl+z to undo it until a page refresh. 

Right now, this matches the behavior of the image block, so it's probably best left as a follow-up to fix in both places.

**Review by vcanales — APPROVED** [review]
> Right now, this matches the behavior of the image block, so it's probably best left as a follow-up to fix in both places.

 Agreed! LGTM :)

**Pull request merged**